CIPs/0009.md

```diff
@@ -0,0 +1,65 @@
+# CIP [0009]: Standalone signing service for Validators
+
+- Date: 2020-05-14
+- Author: @asaj
+- Status: WIP
+
+## Overview
+
+Validator operational excellence is vital to the security and performance of the Celo network.
+
+However, validators currently face significant operational limitations due to key management being bundled into `celo-blockchain`.
+
+In particular, it makes it cumbersome to switch which node is used to participate in consensus.
+
+Switching nodes allows validators to upgrade to newer versions of `celo-blockchain` or to failover to a different node if the primary node fails.
+
+Validators who wish to switch nodes currently have the following options:
+
+1. Bring down node A, bring up node B with key material copied from node A.
+2. Bring up node B with key material copied from node A, bring down node B.
+3. Bring up node B with new key material, authorize the new key, and bring down node A after the next election.
+
+Option `1` results in downtime, option `2` risks double signing, and option `3` can only be done once every 24 hours.
+Furthermore, options `1` and `2` may require unsafe key management practices, which may result in keys being compromised.
+
+Decoupling key management from other validator functionality will allow validators to switch nodes with zero downtime and no risk of double signing.
+
+## Goals
+
+- Zero downtime validator upgrades within an epoch
+- Failover possible with no risk of double signing
+
+## Proposed Solution
+
+The proposed solution is to use `clef`, introduced in [go-ethereum v1.9](https://blog.ethereum.org/2019/07/10/geth-v1-9-0/), as a standalone signer for validators.
+
+`clef` is integrated into `celo-blockchain` as an implementation of the [wallet](https://github.com/celo-org/celo-blockchain/blob/3caca596d1ff23a9faf5339f108c993f38bfa743/accounts/accounts.go#L50) interface in [accounts/external/backend.go](https://github.com/celo-org/celo-blockchain/blob/master/accounts/external/backend.go).
+
+Validators will run a `clef` instance that manages their [validator signer key](https://docs.celo.org/validator-guide/summary/detailed#authorized-validator-signers), and point their node to that instance.
+
+To switch nodes, validators will issue a command to `clef` to stop accepting RPCs from node A and start accepting RPCs from node B.
+
+### Implementation
+
+The current `clef` implementation of the `wallet` interface is missing the following functions needed to support running a validator:
+- Decrypt
```

Comment:
I know this increases scope, it would be great if in the process we could add encrypt and decrypt functionality on the keys, somewhat related to various efforts on Ethereum as well https://github.com/ethereum/EIPs/issues/130

Reply:
That seems orthogonal to me